### PR TITLE
Replace the retry implementation.

### PR DESCRIPTION
**Pull request by nrc**
The changes outside of retry.rs are to support the change from futures to async/await. It is probably easier to review the whole file rather than the diff.

PTAL @sticnarf @Hoverbear

### CODE REVIEW

File: src/pd/retry.rs [hunk 1]
```diff
@@ -131,99 +109,153 @@ impl fmt::Debug for RetryClient {
...
 
-    // A function which makes an async request.
-    func: Func,
+    fn with_cluster<T, F: Fn(&Cluster) -> T>(&self, f: F) -> T {
+        f(&self.cluster.read().unwrap())
+    }
 }
 
-impl<Resp, Fut> Future for RetryRequest<Fut>
+async fn retry_request2<Rc, Resp, Func, RespFuture>(client: Arc<Rc>, func: Func) -> Result<Resp>
```
Comment (sticnarf) [inline src/pd/retry.rs]: Why there is a `2`?

Comment (nrc) [reply src/pd/retry.rs]: whoops, should have taken that out...

File: src/pd/retry.rs [hunk 2]
```diff
@@ -131,99 +109,153 @@ impl fmt::Debug for RetryClient {
...
-    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<Resp>> {
-        unsafe {
-            let this = Pin::get_unchecked_mut(self);
-            if this.reconnect_count == 0 {
-                return Poll::Ready(Err(internal_err!("failed to send request")));
+    let mut last_err = Ok(());
+    for _ in 0..LEADER_CHANGE_RETRY {
+        // Reconnect if necessary.
+        let mut reconnect_count = MAX_REQUEST_COUNT;
+        while let Err(e) = client.reconnect(RECONNECT_INTERVAL_SEC) {
```
Comment (sticnarf) [inline src/pd/retry.rs]: Looks like we always reconnect on the first try. We should reconnect only if there is an error.

Comment (nrc) [reply src/pd/retry.rs]: addressed

**Comment by nrc** [comment]
@sticnarf added another test

**Pull request merged**